### PR TITLE
UI menu

### PR DESCRIPTION
**Pull request by ParthPan7**
Applied Template Design Pattern.

### CODE REVIEW

File: PhysicsExploration/main.cpp [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
+{	
+	int selectedSimulation = 9999;
+	int exitCondition = -9;
+	PhysicsSimulationsFactory* physicsSimulationFactory = new PhysicsSimulationsFactory;
+	 while (exitCondition != selectedSimulation)
+	{
+		RenderingPipeline* renderingPipeline = physicsSimulationFactory->requestPhysicsSimulation(selectedSimulation);
+		selectedSimulation = renderingPipeline->runAppTemplate();
+		delete renderingPipeline;
+	}
```
Comment (RoseDuf) [inline PhysicsExploration/main.cpp:15]: add `delete physicsSimulationFactory`

File: PhysicsExploration/main.cpp [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
 int main()
-{
-	App* application = new App();
-
-	application->runApp();
-
-	delete application;
+{	
+	int selectedSimulation = 9999;
+	int exitCondition = -9;
```
Comment (RoseDuf) [inline PhysicsExploration/main.cpp:8]: why -9 and 9999?

File: PhysicsExploration/RenderingPipeline.cpp [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+	glfwSetCursorPosCallback(_window, cursor_pos_callback_dispatch);
+	glfwSetScrollCallback(_window, scroll_callback_dispatch);
+	//glfwSetInputMode(_window, GLFW_CURSOR, GLFW_CURSOR_DISABLED);
+	//glfwSetMouseButtonCallback(window, mouse_button_callback);
+	//glfwSetInputMode(window, GLFW_STICKY_MOUSE_BUTTONS, 1);
+
+	stbi_set_flip_vertically_on_load(true);
+}
+
+void RenderingPipeline::initializeShaders() const
```
Comment (RoseDuf) [inline PhysicsExploration/RenderingPipeline.cpp:163]: why did you move the implementation of this function to ModelLoading?

File: PhysicsExploration/RenderingPipeline.cpp [hunk 4]
```diff
@@ -0,0 +1,384 @@
+#include "RenderingPipeline.h"
+GLFWwindow* RenderingPipeline::_window;
+float RenderingPipeline::_deltaTime;
+float RenderingPipeline::_lastFrame;
+glm::mat4 RenderingPipeline::_lightSpaceMatrix;
+GLuint RenderingPipeline::_depthMap;
+GLuint RenderingPipeline::_depthMapFBO;
+bool RenderingPipeline::_isMenuEnabled;
+int RenderingPipeline::_selectedItem;
+int RenderingPipeline::_previousItem;
```
Comment (RoseDuf) [inline PhysicsExploration/RenderingPipeline.cpp:10]: why do you re-add the variables here?

File: PhysicsExploration/ModelLoading.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
+#include "ModelLoading.h"
+Shader* RenderingPipeline::_shader;
+Shader* RenderingPipeline::_lightShader;
+Shader* RenderingPipeline::_depthShader;
+Model* RenderingPipeline::_plane;
+Model* RenderingPipeline::_object;
+ModelLoading* ModelLoading::_uniqueInstance = nullptr;
```
Comment (RoseDuf) [inline PhysicsExploration/ModelLoading.cpp:7]: why do you re-add the variables here?

File: PhysicsExploration/ModelLoading.h [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+protected:
+	
+	static ModelLoading* _uniqueInstance;
+public:
+	ModelLoading();
+	~ModelLoading();
+	static ModelLoading* instance();
+	void initializeShaders() const override;
+	void initializeModels() const override;
+	void renderScene(const Shader& shader) const override;
```
Comment (RoseDuf) [inline PhysicsExploration/ModelLoading.h]:
```
void initializeShaders() const override;
void initializeModels() const override;
void renderScene(const Shader& shader) const override;
```
shouldn't all of these be protected too?

File: PhysicsExploration/main.cpp [hunk 7]
```diff
@@ -1,12 +1,17 @@
-#include "App.h"
+#include "ModelLoading.h"
+#include "PhysicsSimulationsFactory.h"
 using namespace std;
 
 int main()
-{
-	App* application = new App();
```
Comment (RoseDuf) [inline PhysicsExploration/main.cpp:6]: Did you delete the App files? I don't see their deletion in the change list

File: PhysicsExploration/RenderingPipeline.h [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	float _lastX;
+	float _lastY;
+	float _fov;
+
+	//engine clock time
+	static float _deltaTime;
+	static float _lastFrame;
+
+	//shader variables
+	const static unsigned int SHADOW_WIDTH = 1024;
```
Comment (RoseDuf) [inline PhysicsExploration/RenderingPipeline.h:54]: why are you making all of these variables static?

File: PhysicsExploration/main.cpp (hunk 1 above)
Comment (ParthPan7) [reply PhysicsExploration/main.cpp:15]: 👍

File: PhysicsExploration/ModelLoading.h (hunk 6 above)
Comment (ParthPan7) [reply PhysicsExploration/ModelLoading.h]: 👍

File: PhysicsExploration/main.cpp (hunk 7 above)
Comment (ParthPan7) [reply PhysicsExploration/main.cpp:6]: I though we have discussion before eliminating it :)

Comment (RoseDuf) [reply PhysicsExploration/main.cpp:6]: You're free to do so!

File: PhysicsExploration/RenderingPipeline.h (hunk 8 above)
Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: It's pattern enforceable - I felt that while I started embedding pattern and way they are being used then after.  If one thinks logically - at any moment of time, only single specific instance overriding RenderingPipeline class is employing it.

Comment (RoseDuf) [reply PhysicsExploration/RenderingPipeline.h:54]: what's the point of having consts static if they never change?

File: PhysicsExploration/main.cpp (hunk 7 above)
Comment (RoseDuf) [reply PhysicsExploration/main.cpp:6]: i saw you deleted App.h but not App.cpp...

File: PhysicsExploration/RenderingPipeline.h (hunk 8 above)
Comment (RoseDuf) [reply PhysicsExploration/RenderingPipeline.h:54]: and what do you mean by pattern enforceable?

Comment (RoseDuf) [reply PhysicsExploration/RenderingPipeline.h:54]: because you make a static instance of an object?

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: I will try to eliminate the consts bearing static if that wouldn't break anything.

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: Let me give you a hint : https://www.reddit.com/r/cpp/comments/x7wxh/why_cant_a_derived_class_access_protected_members/. I am willing to give a lecture on template pattern and what's the difference while it came to applicability in C++ unlike language like JAVA. Do you agree ? I will add my confluence tomorrow or in coming week for your reference.

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: > and what do you mean by pattern enforceable?

Let me give you a hint : https://www.reddit.com/r/cpp/comments/x7wxh/why_cant_a_derived_class_access_protected_members/. I am willing to give a lecture on template pattern and what's the difference when it comes to applicability in C++ unlike language like JAVA. Do you agree ? I will add my confluence tomorrow or in coming week for your reference.

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: I need to conduct lecture for this thread! or Q&A I think.

Comment (RoseDuf) [reply PhysicsExploration/RenderingPipeline.h:54]: in what context does our code have a similar case than the example shown in that link?
```
class Base
{
protected:
  void DoSomething() {}
};

class Derived : public Base
{
public:
  void Test() 
  {
    Base base_class;
    base_class.DoSomething();
  }
};

int main(int argc, char* argv[])
{
  Derived der;
  der.Test();
  return 0;
}
```

Do we have a case where we have a function like "Test()" that creates an object of its own parent class?

if the base class's function is protected, couldn't "Test()" just do:
```
void Test() 
  {
    DoSomething();
  }
```

since it inherits the base class function already anyways?

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: C++ uses some fairly odd scoped inheritance rules. When you have a protected member, you mean that instances deriving from the base are allowed free access to the field in this object. You are not saying that derivatives of the base are allowed to access any instance (which might not even be their own type). Consider:

class Base
{
protected:
    int x;
};

class Derived : public Base { };

class Sneaker : public Base
{
public:
    static void AlterX(Derived& d)
    {
        // One might think that I can access x here, because it comes from Base
        // and Sneaker inherits from Base; however, no regular casting will
        // allow me to do so (there is probably some incantation that does).
        d.x = 10;
    }
};

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: > C++ uses some fairly odd scoped inheritance rules. When you have a protected member, you mean that instances deriving from the base are allowed free access to the field in this object. You are not saying that derivatives of the base are allowed to access any instance (which might not even be their own type). Consider:
> 
> class Base { protected: int x; };
> 
> class Derived : public Base { };
> 
> class Sneaker : public Base { public: static void AlterX(Derived& d) { // One might think that I can access x here, because it comes from Base // and Sneaker inherits from Base; however, no regular casting will // allow me to do so (there is probably some incantation that does). d.x = 10; } };

This was the case I was facing. You can't access protected 'x' in Sneaker class if you want directly. This is the reply from the same reddit thread :)

Comment (RoseDuf) [reply PhysicsExploration/RenderingPipeline.h:54]: I see, that makes sense, but are you doing this type of thing anywhere?
In what case would a derived class need a reference to another derived class in our case?

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: Nope, just try to convince yourself that protected member 'x' of the base class can't be accessed from the inherited class whatsoever. What they have portrayed is a way of accessing it - refer that comment inside the code Rose : )
Template is designed such that it needs direct access of protected data members of the Base class.

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: > Nope, just try to convince yourself that protected member 'x' of the base class can't be accessed from the inherited class whatsoever. What they have portrayed is a way of accessing it - refer that comment inside the code Rose : ) Template is designed such that it needs direct access of protected data members of the Base class.

// One might think that I can access x here, because it comes from Base
// and Sneaker inherits from Base; however, no regular casting will
// allow me to do so (there is probably some incantation that does).

Comment (RoseDuf) [reply PhysicsExploration/RenderingPipeline.h:54]: That isn't true though. derived classes DO get access to the protected variables of the base class in normal circumstances. The only reason why you aren't able to access them in this code you made is because you made all of the functions "const" and consts are implicitly static, which means they can only use static variables within them...

I don't understand why you made all those functions const in the first place

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: Thanks for letting me know. I think I have been taking this documentation as a standard one : https://refactoring.guru/design-patterns/template-method/cpp/example. One way of doing this pattern - somehow during implementation of this pattern it's blocking data member access in child class.

Comment (RoseDuf) [reply PhysicsExploration/RenderingPipeline.h:54]: ahh ok, this makes a lot of sense. If this is the case then I guess everything is ok! :)

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: If you feel the way I refactored is unsafe I can remove const and revised the code version as per.

Comment (RoseDuf) [reply PhysicsExploration/RenderingPipeline.h:54]: it should be good. i just needed more context, buit i understand now

Comment (ParthPan7) [reply PhysicsExploration/RenderingPipeline.h:54]: I am revisting c++  concepts :P

Comment (RoseDuf) [reply PhysicsExploration/RenderingPipeline.h:54]: me too haha
